libraries/stumbler/src/main/java/org/mozilla/mozstumbler/service/core/offline/LocationService.java

```diff
@@ -0,0 +1,315 @@
+                // and use that as an average to figure out which single tile
+                // we should return
+                Log.i(LOG_TAG, "Multiple solutions- take the average: " + maxpt_tileset);
+
+                Set<Integer> tiebreaking_set = new HashSet<Integer>();
+                List<Integer> maxpt_tilelist = new ArrayList<Integer>(maxpt_tileset);
+                for (int i = 0; i < maxpt_tilelist.size(); i++) {
+                    Integer pt = maxpt_tilelist.get(i);
+
+                    List<Integer> adjacent_tilelist = adjacent_tile(pt);
+                    HashSet<Integer> adjacent_tileset = new HashSet<Integer>(adjacent_tilelist);
+                    adjacent_tileset.retainAll(new HashSet<Integer>(maxpt_tilelist.subList(i+1, maxpt_tilelist.size())));
+                    if (adjacent_tileset.size() > 0) {
+                        tiebreaking_set.add(pt);
+                    }
+                }
+                double final_lat = 0;
+                double final_lon = 0;
+                for (Integer tie_tileid: tiebreaking_set) {
+                    SmartTile tmp_coord = city_tiles.getCoord(tie_tileid);
+                    final_lat += (tmp_coord.getLocation().getLatitude() * 1.0 / tiebreaking_set.size());
+                    final_lon += (tmp_coord.getLocation().getLongitude() * 1.0 / tiebreaking_set.size());
+                }
+
+                SmartTile final_coord = new SmartTile(final_lat, final_lon);
+                Log.i(LOG_TAG, "Multiple solutions converging on : " + final_coord);
+                return locationFix(city_tiles.getTileID(final_coord));
+            }
+        }
+
+        Log.i(LOG_TAG, "Can't find a solution");
+        return null;
+    }
+
+    private List<Integer> adjacent_tile(int tile_id) {
+        ArrayList<Integer> result = new ArrayList<Integer>();
+
+        SmartTile tc = city_tiles.getCoord(tile_id);
+        if (tc == null) {
+            // No adjacent tiles could be found.
+            Log.w(LOG_TAG, "Couldn't find adjacent tiles for tile_id=[" + tile_id + "]");
+            return result;
+        }
+        safe_add_adjacent_tile(city_tiles, result, tc.tile_x - 1, tc.tile_y - 1);
+        safe_add_adjacent_tile(city_tiles, result, tc.tile_x, tc.tile_y - 1);
+        safe_add_adjacent_tile(city_tiles, result, tc.tile_x + 1, tc.tile_y - 1);
+
+        safe_add_adjacent_tile(city_tiles, result, tc.tile_x - 1, tc.tile_y);
+        safe_add_adjacent_tile(city_tiles, result, tc.tile_x + 1, tc.tile_y);
+
+        safe_add_adjacent_tile(city_tiles, result, tc.tile_x - 1, tc.tile_y + 1);
+        safe_add_adjacent_tile(city_tiles, result, tc.tile_x, tc.tile_y + 1);
+        safe_add_adjacent_tile(city_tiles, result, tc.tile_x + 1, tc.tile_y + 1);
+
+        return result;
+    }
+
+    private void safe_add_adjacent_tile(OrderedCityTiles city_tiles, ArrayList<Integer> result, int tile_x, int tile_y) {
+        int tile_id = city_tiles.getTileID(new SmartTile(tile_x, tile_y));
```

Comment:
This method only exists for the edges of a city.  Edge tiles will have < 8 adjacent tiles. Make a comment here for future.